rafthttp/snapshot_sender.go

```diff
@@ -0,0 +1,205 @@
+		s.status.activate()
+		s.r.ReportSnapshot(m.To, raft.SnapshotFinish)
+		plog.Infof("snapshot [index: %d, to: %s] sent out successfully", m.Snapshot.Metadata.Index, types.ID(m.To))
+	}
+}
+
+// post posts the given request.
+// It returns nil when request is sent out and processed successfully.
+func (s *snapshotSender) post(req *http.Request) (err error) {
+	cancel := httputil.RequestCanceler(s.tr, req)
+
+	type responseAndError struct {
+		resp *http.Response
+		body []byte
+		err  error
+	}
+	result := make(chan responseAndError, 1)
+
+	go func() {
+		// TODO: cancel the request if it has waited for a long time(~5s) after
+		// it has write out the full request body, which helps to avoid receiver
+		// dies when sender is waiting for response
+		resp, err := s.tr.RoundTrip(req)
+		if err != nil {
+			result <- responseAndError{resp, nil, err}
+			return
+		}
+		body, err := ioutil.ReadAll(resp.Body)
+		defer resp.Body.Close()
+		result <- responseAndError{resp, body, err}
+	}()
+
+	var resp *http.Response
+	var body []byte
+	select {
+	case <-s.stopc:
+		cancel()
+		return errStopped
+	case r := <-result:
+		if r.err != nil {
+			return r.err
+		}
+		resp, body = r.resp, r.body
+	}
+
+	switch resp.StatusCode {
+	case http.StatusPreconditionFailed:
+		switch strings.TrimSuffix(string(body), "\n") {
+		case errIncompatibleVersion.Error():
+			plog.Errorf("request sent was ignored by peer %s (server version incompatible)", s.to)
+			return errIncompatibleVersion
+		case errClusterIDMismatch.Error():
+			plog.Errorf("request sent was ignored (cluster ID mismatch: remote[%s]=%s, local=%s)",
+				s.to, resp.Header.Get("X-Etcd-Cluster-ID"), s.cid)
+			return errClusterIDMismatch
+		default:
+			return fmt.Errorf("unhandled error %q when precondition failed", string(body))
+		}
+	case http.StatusForbidden:
+		err := fmt.Errorf("the member has been permanently removed from the cluster")
+		select {
+		case s.errorc <- err:
+		default:
+		}
+		return err
+	case http.StatusNoContent:
+		return nil
+	default:
+		return fmt.Errorf("unexpected http status %s while posting to %q", http.StatusText(resp.StatusCode), req.URL.String())
+	}
+}
+
+// readCloser implements io.ReadCloser interface.
+type readCloser struct {
+	io.Reader
+	io.Closer
+}
+
+// createSnapBody creates the request body for the given raft snapshot message.
+// The returned body should always be closed when it is no longer used.
+func createSnapBody(m raftpb.Message, snapst *snapshotStore) io.ReadCloser {
+	buf := new(bytes.Buffer)
+	// put the length of raft message into the buffer
+	if err := binary.Write(buf, binary.BigEndian, uint64(m.Size())); err != nil {
+		plog.Panicf("binary write error (%v)", err)
+	}
+	// put the raft message into the buffer
+	if _, err := buf.Write(pbutil.MustMarshal(&m)); err != nil {
+		plog.Panicf("body write error (%v)", err)
+	}
+	// get snapshot
+	// TODO: the snapshot could be large and eat up all resources when writing
```

Comment:
Basically the new Reader should be a kind-of `PartTimeReader` upon original `io.Reader`. It is better to add this layer in send() function. So I will remove it from building the body.